File: artwork_validator/core/enhanced_validator.py
# core/enhanced_validator.py
import re
import logging
from typing import List, Dict, Tuple, Optional, Set
from difflib import SequenceMatcher

class EnhancedValidator:
    """
    Validateur amélioré avec correspondance séquentielle et élimination progressive
    pour éviter les correspondances multiples et assurer une validation 1:1
    """

    def __init__(self):
        self.equivalences = {
            'WTP': 'WATERPROOF',
            'WATERPROOF': 'WTP'
        }
        self.check_digits = False
        self.logger = logging.getLogger(__name__)

        # Métriques de validation
        self.validation_stats = {
            'total_tokens': 0,
            'consumed_tokens': 0,
            'orphan_tokens': 0,
            'exact_matches': 0,
            'partial_matches': 0,
            'failed_matches': 0,
            'duplicate_allowances': 0
        }

    def _safe_str(self, value) -> str:
        """Convertit de manière sécurisée une valeur en string"""
        if value is None:
            return ""
        if isinstance(value, (int, float)):
            if float(value).is_integer():
                return str(int(value))
            return str(value)
        return str(value).strip()

    def _tokenize_pdf_text(self, pdf_text: str) -> List[Dict]:
        """
        Tokenise le texte PDF en gardant la position et les métadonnées
        """
        # Normaliser le texte
        text = pdf_text.upper().strip()

        # Diviser en tokens (mots/nombres/symboles)
        # Pattern pour capturer mots, nombres, et symboles significatifs
        pattern = r'\b\w+\b|\d+|[A-Z]+'
        tokens = []

        for match in re.finditer(pattern, text):
            token = {
                'text': match.group(),
                'start_pos': match.start(),
                'end_pos': match.end(),
                'consumed': False,
                'consumed_by': None,  # Quelle entrée Excel a consommé ce token
                'type': self._classify_token(match.group())
            }
            tokens.append(token)

        self.validation_stats['total_tokens'] = len(tokens)
        self.logger.info(f"📝 PDF tokenisé: {len(tokens)} tokens extraits")

        return tokens

    def _classify_token(self, token: str) -> str:
        """Classifie le type de token pour optimiser la recherche"""
        if token.isdigit():
            if len(token) == 4:
                return 'digits_4'
            elif len(token) in [2, 3]:
                return 'shade_number'
            else:
                return 'number'
        elif token.isalpha():
            if token in ['WATERPROOF', 'WTP']:
                return 'waterproof_tag'
            elif len(token) > 1:
                return 'shade_name'
            else:
                return 'letter'
        else:
            return 'mixed'

    def _normalize_excel_data(self, excel_data: List[Dict]) -> List[Dict]:
        """
        Normalise les données Excel pour la correspondance
        """
        normalized = []

        for row in excel_data:
            normalized_row = row.copy()

            # Identifier les types spéciaux
            facing_value = self._safe_str(row.get('PRODUCT FACING SL'))
            is_frame = facing_value == 'FRAME'
            is_space_saver = any(
                self._safe_str(row.get(field)) == 'SPACE_SAVER'
                for field in ['UPC', 'PRODUCT DESCRIPTION', 'SHADE NAME']
            )

            # Normaliser les noms de teintes WTP -> WATERPROOF
            shade_name = self._safe_str(row.get('SHADE NAME'))
            if shade_name and 'WTP' in shade_name:
                normalized_row['SHADE_NAME_NORMALIZED'] = shade_name.replace('WTP', 'WATERPROOF')
            else:
                normalized_row['SHADE_NAME_NORMALIZED'] = shade_name

            # Métadonnées pour le traitement
            normalized_row['_is_frame'] = is_frame
            normalized_row['_is_space_saver'] = is_space_saver
            normalized_row['_should_validate'] = not (is_frame or is_space_saver)

            normalized.append(normalized_row)

        return normalized

    def _find_token_match(self, tokens: List[Dict], search_text: str,
                         start_index: int = 0, token_types: List[str] = None) -> Optional[Tuple[int, Dict, int]]:
        """
        Trouve la première correspondance d'un texte dans les tokens à partir d'un index.
        Retourne: (index_début, premier_token, nombre_de_tokens_consommés)
        """
        if not search_text:
            return None

        search_text = search_text.upper().strip()
        search_words = search_text.split()  # Diviser en mots pour les noms composés

        # CAS 1: Recherche multi-tokens (pour les noms composés comme "FOREST BROWN")
        if len(search_words) > 1:
            for i in range(start_index, len(tokens) - len(search_words) + 1):
                # Vérifier si les tokens ne sont pas déjà consommés
                tokens_slice = tokens[i:i + len(search_words)]
                if any(t['consumed'] for t in tokens_slice):
                    continue

                # Vérifier si les tokens correspondent aux mots
                match = True
                for j, word in enumerate(search_words):
                    if tokens[i + j]['text'] != word:
                        match = False
                        break

                if match:
                    # Retourner l'index de début, le premier token, et le nombre de tokens
                    return i, tokens[i], len(search_words)

        # CAS 2: Recherche single-token (comportement original)
        for i in range(start_index, len(tokens)):
            token = tokens[i]

            # Ignorer les tokens déjà consommés
            if token['consumed']:
                continue

            # Filtrer par type si spécifié
            if token_types and token['type'] not in token_types:
                continue

            # Correspondance exacte
            if token['text'] == search_text:
                return i, token, 1

            # Correspondance partielle pour les noms longs
            if len(search_text) > 3 and search_text in token['text']:
                return i, token, 1

        return None

    def _consume_token(self, tokens: List[Dict], token_index: int, consumed_by: str, num_tokens: int = 1):
        """
        Marque un ou plusieurs tokens consécutifs comme consommés

        Args:
            tokens: Liste des tokens
            token_index: Index de départ
            consumed_by: Identifiant de l'entrée qui consomme
            num_tokens: Nombre de tokens à consommer (défaut: 1)
        """
        for i in range(token_index, min(token_index + num_tokens, len(tokens))):
            if 0 <= i < len(tokens):
                tokens[i]['consumed'] = True
                tokens[i]['consumed_by'] = consumed_by
                self.validation_stats['consumed_tokens'] += 1

    def _check_duplicate_allowance(self, excel_data: List[Dict], current_index: int) -> bool:
        """
        Vérifie si l'entrée courante est un doublon légitime
        """
        if current_index == 0:
            return False

        current_row = excel_data[current_index]

        # Comparer avec les entrées précédentes
        for i in range(current_index):
            prev_row = excel_data[i]

            # Même shade number et shade name = doublon légitime
            if (self._safe_str(current_row.get('SHADE NUMBER')) ==
                self._safe_str(prev_row.get('SHADE NUMBER')) and
                self._safe_str(current_row.get('SHADE NAME')) ==
                self._safe_str(prev_row.get('SHADE NAME'))):
                return True

        return False

    def validate_enhanced(self, pdf_text: str, excel_data: List[Dict]) -> Dict:
        """
        Validation améliorée avec correspondance séquentielle et élimination progressive
        """
        self.logger.info("🔍 Début de la validation améliorée")

        # Réinitialiser les statistiques
        self.validation_stats = {key: 0 for key in self.validation_stats}

        if not excel_data or not pdf_text:
            return {'results': [], 'stats': self.validation_stats, 'errors': ['Données manquantes']}

        # 1. Tokeniser le PDF
        tokens = self._tokenize_pdf_text(pdf_text)

        # 2. Normaliser les données Excel
        normalized_excel = self._normalize_excel_data(excel_data)

        # 3. Validation séquentielle
        results = []
        current_position = 0
        errors = []

        for idx, row in enumerate(normalized_excel):
            result = self._validate_single_entry(tokens, row, idx, current_position)
            results.append(result)

            # Mettre à jour la position pour la prochaine recherche
            if result.get('last_token_position') is not None:
                current_position = result['last_token_position'] + 1

        # 4. Analyser les tokens orphelins
        orphan_tokens = [t for t in tokens if not t['consumed']]
        self.validation_stats['orphan_tokens'] = len(orphan_tokens)

        if orphan_tokens:
            self.logger.warning(f"⚠️ {len(orphan_tokens)} tokens non consommés détectés")
            for token in orphan_tokens[:5]:  # Limiter le log
                self.logger.warning(f"   Token orphelin: '{token['text']}'")

        # 5. Vérification du décompte
        expected_validations = sum(1 for row in normalized_excel if row['_should_validate'])
        actual_validations = sum(1 for r in results if r.get('overall', False))

        if expected_validations != actual_validations:
            errors.append(f"Décompte incorrect: {actual_validations}/{expected_validations} validations")

        self.logger.info(f"✅ Validation améliorée terminée: {actual_validations}/{expected_validations} succès")

        return {
            'results': results,
            'stats': self.validation_stats,
            'errors': errors,
            'orphan_tokens': [t['text'] for t in orphan_tokens],
            'summary': {
                'expected_validations': expected_validations,
                'actual_validations': actual_validations,
                'success_rate': (actual_validations / expected_validations * 100) if expected_validations > 0 else 0
            }
        }

    def _validate_single_entry(self, tokens: List[Dict], row: Dict, row_index: int, start_position: int) -> Dict:
        """
        Valide une entrée Excel individuelle contre les tokens PDF
        """
        entry_id = f"Row_{row_index + 1}"

        result = {
            'shade_number': True,
            'shade_name': True,
            'digits': True,
            'facing': row.get('PRODUCT FACING SL', ''),
            'is_mixed': False,  # À déterminer par le processus principal
            'is_cubby': False,
            'is_frame': row['_is_frame'],
            'is_space_saver': row['_is_space_saver'],
            'overall': True,
            'validation_details': {
                'method': 'enhanced',
                'tokens_used': [],
                'start_position': start_position,
                'last_token_position': None
            }
        }

        # Ne pas valider les FRAME et SPACE_SAVER
        if not row['_should_validate']:
            return result

        # Vérifier si c'est un doublon légitime
        is_duplicate = self._check_duplicate_allowance([row] + [{}] * row_index, row_index)
        if is_duplicate:
            self.validation_stats['duplicate_allowances'] += 1

        current_pos = start_position
        validation_success = True

        # 1. Validation du shade number
        shade_number = self._safe_str(row.get('SHADE NUMBER'))
        if shade_number:
            match = self._find_token_match(tokens, shade_number, current_pos, ['shade_number', 'number'])
            if match:
                token_idx, token, num_tokens = match  # ⭐ Récupérer le nombre de tokens
                if not is_duplicate:  # Seulement consommer si pas un doublon
                    self._consume_token(tokens, token_idx, f"{entry_id}_SHADE_NUMBER", num_tokens)
                result['validation_details']['tokens_used'].append({
                    'field': 'SHADE_NUMBER',
                    'token': token['text'],
                    'position': token_idx,
                    'num_tokens': num_tokens
                })
                current_pos = token_idx + num_tokens  # ⭐ Avancer de num_tokens
                result['validation_details']['last_token_position'] = token_idx + num_tokens - 1
                self.validation_stats['exact_matches'] += 1
            else:
                result['shade_number'] = False
                validation_success = False
                self.validation_stats['failed_matches'] += 1

        # 2. Validation du shade name
        shade_name = row.get('SHADE_NAME_NORMALIZED', '')
        if shade_name:
            match = self._find_token_match(tokens, shade_name, current_pos, ['shade_name', 'mixed'])
            if match:
                token_idx, token, num_tokens = match  # ⭐ Récupérer le nombre de tokens
                if not is_duplicate:
                    self._consume_token(tokens, token_idx, f"{entry_id}_SHADE_NAME", num_tokens)
                result['validation_details']['tokens_used'].append({
                    'field': 'SHADE_NAME',
                    'token': token['text'],
                    'position': token_idx,
                    'num_tokens': num_tokens
                })
                current_pos = token_idx + num_tokens  # ⭐ Avancer de num_tokens
                result['validation_details']['last_token_position'] = token_idx + num_tokens - 1
                self.validation_stats['exact_matches'] += 1
            else:
                result['shade_name'] = False
                validation_success = False
                self.validation_stats['failed_matches'] += 1

        # 3. Validation des 4 digits (si activée)
        if self.check_digits:
            digits = self._safe_str(row.get('4 DIGITS'))
            if digits:
                match = self._find_token_match(tokens, digits, current_pos, ['digits_4', 'number'])
                if match:
                    token_idx, token, num_tokens = match  # ⭐ Récupérer le nombre de tokens
                    if not is_duplicate:
                        self._consume_token(tokens, token_idx, f"{entry_id}_4DIGITS", num_tokens)
                    result['validation_details']['tokens_used'].append({
                        'field': '4_DIGITS',
                        'token': token['text'],
                        'position': token_idx,
                        'num_tokens': num_tokens
                    })
                    current_pos = token_idx + num_tokens  # ⭐ Avancer de num_tokens
                    result['validation_details']['last_token_position'] = token_idx + num_tokens - 1
                    self.validation_stats['exact_matches'] += 1
                else:
                    result['digits'] = False
                    validation_success = False
                    self.validation_stats['failed_matches'] += 1

        # Validation globale
        validation_criteria = [result['shade_number'], result['shade_name']]
        if self.check_digits and self._safe_str(row.get('4 DIGITS')):
            validation_criteria.append(result['digits'])

        result['overall'] = all(validation_criteria)

        return result

File: artwork-validator-web/src/core/enhancedValidator.ts
// Port of core/enhanced_validator.py — token-based sequential validation with
// progressive token consumption for 1:1 matching.
import { safeStr } from './textSafe'
import type { LithoRecord } from './types'

export interface PdfToken {
  text: string
  start_pos: number
  end_pos: number
  consumed: boolean
  consumed_by: string | null
  type: TokenType
}

export type TokenType =
  | 'digits_4'
  | 'shade_number'
  | 'number'
  | 'waterproof_tag'
  | 'shade_name'
  | 'letter'
  | 'mixed'

export interface TokenUsage {
  field: string
  token: string
  position: number
  num_tokens: number
}

export interface EnhancedEntryResult {
  shade_number: boolean
  shade_name: boolean
  digits: boolean
  facing: string | number
  is_mixed: boolean
  is_cubby: boolean
  is_frame: boolean
  is_space_saver: boolean
  overall: boolean
  validation_details: {
    method: 'enhanced'
    tokens_used: TokenUsage[]
    start_position: number
    last_token_position: number | null
  }
}

export interface ValidationStats {
  total_tokens: number
  consumed_tokens: number
  orphan_tokens: number
  exact_matches: number
  partial_matches: number
  failed_matches: number
  duplicate_allowances: number
}

export interface EnhancedValidationResult {
  results: EnhancedEntryResult[]
  stats: ValidationStats
  errors: string[]
  orphan_tokens?: string[]
  summary?: {
    expected_validations: number
    actual_validations: number
    success_rate: number
  }
}

interface NormalizedRow {
  row: LithoRecord
  SHADE_NAME_NORMALIZED: string
  _is_frame: boolean
  _is_space_saver: boolean
  _should_validate: boolean
}

/** Python str.isdigit() approximation for our uppercase-ASCII token stream. */
function isDigits(s: string): boolean {
  return /^[0-9]+$/.test(s)
}

/** Python str.isalpha() — Unicode letters only. */
function isAlpha(s: string): boolean {
  return /^\p{L}+$/u.test(s)
}

export class EnhancedValidator {
  checkDigits = false
  validationStats: ValidationStats = this.freshStats()

  private freshStats(): ValidationStats {
    return {
      total_tokens: 0,
      consumed_tokens: 0,
      orphan_tokens: 0,
      exact_matches: 0,
      partial_matches: 0,
      failed_matches: 0,
      duplicate_allowances: 0,
    }
  }

  /**
   * Tokenizes the PDF text. Python uses `\b\w+\b|\d+|[A-Z]+` where `\w` is
   * Unicode-aware; the JS equivalent is `[\p{L}\p{N}_]+` so accented shade
   * names (ROSÉ) stay one token like in Python.
   */
  tokenizePdfText(pdfText: string): PdfToken[] {
    const text = pdfText.toUpperCase().trim()
    const tokens: PdfToken[] = []
    const pattern = /[\p{L}\p{N}_]+/gu
    let match: RegExpExecArray | null
    while ((match = pattern.exec(text)) !== null) {
      tokens.push({
        text: match[0],
        start_pos: match.index,
        end_pos: match.index + match[0].length,
        consumed: false,
        consumed_by: null,
        type: this.classifyToken(match[0]),
      })
    }
    this.validationStats.total_tokens = tokens.length
    return tokens
  }

  classifyToken(token: string): TokenType {
    if (isDigits(token)) {
      if (token.length === 4) return 'digits_4'
      if (token.length === 2 || token.length === 3) return 'shade_number'
      return 'number'
    }
    if (isAlpha(token)) {
      if (token === 'WATERPROOF' || token === 'WTP') return 'waterproof_tag'
      if (token.length > 1) return 'shade_name'
      return 'letter'
    }
    return 'mixed'
  }

  private normalizeExcelData(excelData: LithoRecord[]): NormalizedRow[] {
    return excelData.map((row) => {
      const facingValue = safeStr(row['PRODUCT FACING SL'])
      const isFrame = facingValue === 'FRAME'
      const isSpaceSaver = ['UPC', 'PRODUCT DESCRIPTION', 'SHADE NAME'].some(
        (field) => safeStr(row[field]) === 'SPACE_SAVER',
      )
      const shadeName = safeStr(row['SHADE NAME'])
      const normalized =
        shadeName && shadeName.includes('WTP')
          ? shadeName.replaceAll('WTP', 'WATERPROOF')
          : shadeName
      return {
        row,
        SHADE_NAME_NORMALIZED: normalized,
        _is_frame: isFrame,
        _is_space_saver: isSpaceSaver,
        _should_validate: !(isFrame || isSpaceSaver),
      }
    })
  }

  /**
   * Finds the first match of `searchText` in the tokens starting at `startIndex`.
   * Returns [startIdx, firstToken, numTokensConsumed] or null.
   */
  findTokenMatch(
    tokens: PdfToken[],
    searchText: string,
    startIndex = 0,
    tokenTypes: TokenType[] | null = null,
  ): [number, PdfToken, number] | null {
    if (!searchText) return null

    const search = searchText.toUpperCase().trim()
    const searchWords = search.split(/\s+/).filter((w) => w.length > 0)

    // CASE 1: multi-token search (compound names like "FOREST BROWN")
    if (searchWords.length > 1) {
      for (let i = startIndex; i <= tokens.length - searchWords.length; i++) {
        const slice = tokens.slice(i, i + searchWords.length)
        if (slice.some((t) => t.consumed)) continue
        let match = true
        for (let j = 0; j < searchWords.length; j++) {
          if (tokens[i + j].text !== searchWords[j]) {
            match = false
            break
          }
        }
        if (match) return [i, tokens[i], searchWords.length]
      }
    }

    // CASE 2: single-token search
    for (let i = startIndex; i < tokens.length; i++) {
      const token = tokens[i]
      if (token.consumed) continue
      if (tokenTypes && !tokenTypes.includes(token.type)) continue
      if (token.text === search) return [i, token, 1]
      // Partial match for long names
      if (search.length > 3 && token.text.includes(search)) return [i, token, 1]
    }

    return null
  }

  private consumeToken(tokens: PdfToken[], tokenIndex: number, consumedBy: string, numTokens = 1): void {
    for (let i = tokenIndex; i < Math.min(tokenIndex + numTokens, tokens.length); i++) {
      if (i >= 0 && i < tokens.length) {
        tokens[i].consumed = true
        tokens[i].consumed_by = consumedBy
        this.validationStats.consumed_tokens += 1
      }
    }
  }

  /** Same-shade-number + same-shade-name rows count as legitimate duplicates. */
  checkDuplicateAllowance(excelData: LithoRecord[], currentIndex: number): boolean {
    if (currentIndex === 0) return false
    const currentRow = excelData[currentIndex]
    for (let i = 0; i < currentIndex; i++) {
      const prevRow = excelData[i]
      if (
        safeStr(currentRow['SHADE NUMBER']) === safeStr(prevRow['SHADE NUMBER']) &&
        safeStr(currentRow['SHADE NAME']) === safeStr(prevRow['SHADE NAME'])
      ) {
        return true
      }
    }
    return false
  }

  validateEnhanced(pdfText: string, excelData: LithoRecord[]): EnhancedValidationResult {
    this.validationStats = this.freshStats()

    if (!excelData.length || !pdfText) {
      return { results: [], stats: this.validationStats, errors: ['Données manquantes'] }
    }

    const tokens = this.tokenizePdfText(pdfText)
    const normalizedExcel = this.normalizeExcelData(excelData)

    const results: EnhancedEntryResult[] = []
    let currentPosition = 0
    const errors: string[] = []

    normalizedExcel.forEach((row, idx) => {
      const result = this.validateSingleEntry(tokens, row, idx, currentPosition)
      results.push(result)
      // PARITY: reproduces a Python bug — validate_enhanced reads
      // `result.get('last_token_position')` at the TOP level of the dict, but
      // the key only exists inside validation_details, so the cross-row
      // position never advances (every row starts searching at 0). Token
      // consumption still prevents double-matching.
    })

    const orphanTokens = tokens.filter((t) => !t.consumed)
    this.validationStats.orphan_tokens = orphanTokens.length

    const expectedValidations = normalizedExcel.filter((r) => r._should_validate).length
    const actualValidations = results.filter((r) => r.overall).length
    if (expectedValidations !== actualValidations) {
      errors.push(`Décompte incorrect: ${actualValidations}/${expectedValidations} validations`)
    }

    return {
      results,
      stats: this.validationStats,
      errors,
      orphan_tokens: orphanTokens.map((t) => t.text),
      summary: {
        expected_validations: expectedValidations,
        actual_validations: actualValidations,
        success_rate: expectedValidations > 0 ? (actualValidations / expectedValidations) * 100 : 0,
      },
    }
  }

  private validateSingleEntry(
    tokens: PdfToken[],
    row: NormalizedRow,
    rowIndex: number,
    startPosition: number,
  ): EnhancedEntryResult {
    const entryId = `Row_${rowIndex + 1}`

    const result: EnhancedEntryResult = {
      shade_number: true,
      shade_name: true,
      digits: true,
      facing: (row.row['PRODUCT FACING SL'] ?? '') as string | number,
      is_mixed: false,
      is_cubby: false,
      is_frame: row._is_frame,
      is_space_saver: row._is_space_saver,
      overall: true,
      validation_details: {
        method: 'enhanced',
        tokens_used: [],
        start_position: startPosition,
        last_token_position: null,
      },
    }

    if (!row._should_validate) return result

    // PARITY: reproduces a Python bug — the duplicate check receives
    // `[row] + [{}] * row_index`, so every row with index >= 2 is treated as a
    // legitimate duplicate ('' == ''). Kept 1:1 so web results match desktop.
    const paddedData: LithoRecord[] = [row.row, ...Array.from({ length: rowIndex }, () => ({}) as LithoRecord)]
    const isDuplicate = this.checkDuplicateAllowance(paddedData, rowIndex)
    if (isDuplicate) this.validationStats.duplicate_allowances += 1

    let currentPos = startPosition

    // 1. Shade number
    const shadeNumber = safeStr(row.row['SHADE NUMBER'])
    if (shadeNumber) {
      const match = this.findTokenMatch(tokens, shadeNumber, currentPos, ['shade_number', 'number'])
      if (match) {
        const [tokenIdx, token, numTokens] = match
        if (!isDuplicate) this.consumeToken(tokens, tokenIdx, `${entryId}_SHADE_NUMBER`, numTokens)
        result.validation_details.tokens_used.push({
          field: 'SHADE_NUMBER',
          token: token.text,
          position: tokenIdx,
          num_tokens: numTokens,
        })
        currentPos = tokenIdx + numTokens
        result.validation_details.last_token_position = tokenIdx + numTokens - 1
        this.validationStats.exact_matches += 1
      } else {
        result.shade_number = false
        this.validationStats.failed_matches += 1
      }
    }

    // 2. Shade name
    const shadeName = row.SHADE_NAME_NORMALIZED
    if (shadeName) {
      const match = this.findTokenMatch(tokens, shadeName, currentPos, ['shade_name', 'mixed'])
      if (match) {
        const [tokenIdx, token, numTokens] = match
        if (!isDuplicate) this.consumeToken(tokens, tokenIdx, `${entryId}_SHADE_NAME`, numTokens)
        result.validation_details.tokens_used.push({
          field: 'SHADE_NAME',
          token: token.text,
          position: tokenIdx,
          num_tokens: numTokens,
        })
        currentPos = tokenIdx + numTokens
        result.validation_details.last_token_position = tokenIdx + numTokens - 1
        this.validationStats.exact_matches += 1
      } else {
        result.shade_name = false
        this.validationStats.failed_matches += 1
      }
    }

    // 3. 4 DIGITS (when enabled)
    if (this.checkDigits) {
      const digits = safeStr(row.row['4 DIGITS'])
      if (digits) {
        const match = this.findTokenMatch(tokens, digits, currentPos, ['digits_4', 'number'])
        if (match) {
          const [tokenIdx, token, numTokens] = match
          if (!isDuplicate) this.consumeToken(tokens, tokenIdx, `${entryId}_4DIGITS`, numTokens)
          result.validation_details.tokens_used.push({
            field: '4_DIGITS',
            token: token.text,
            position: tokenIdx,
            num_tokens: numTokens,
          })
          currentPos = tokenIdx + numTokens
          result.validation_details.last_token_position = tokenIdx + numTokens - 1
          this.validationStats.exact_matches += 1
        } else {
          result.digits = false
          this.validationStats.failed_matches += 1
        }
      }
    }

    const criteria = [result.shade_number, result.shade_name]
    if (this.checkDigits && safeStr(row.row['4 DIGITS'])) {
      criteria.push(result.digits)
    }
    result.overall = criteria.every(Boolean)

    return result
  }
}
